realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -1048,6 +1081,151 @@ void BaseRealSenseNode::publishStaticTransforms()
     }
 }
 
+/* Modified for Brain */
+static void output_unit_vector(unsigned int height, unsigned int width, float unitVectors[FRAME_MAX_HEIGHT][FRAME_MAX_WIDTH * 3])
+{
+	unsigned int i, j;
+	FILE *fp; /* output to .obj file with a certain format to be shown by some tool */
+
+	std::cerr << "\nUnit Test for Realsense Unit Vector:\nOutput the saved data: frame width is " \
+		<< width << " and frame height is " << height << " \n\n" << std::endl;
+
+	std::cerr << "*** Beginning of unit vectors ***\n" << std::endl;
+
+	fp = fopen("./realsense_unit_vectors.obj","w");
+	if (!fp) {
+		std::cerr << "\nFailed to open a file for realsense_unit_vectors.obj!\n\n" << std::endl;
+		return;
+	}
+
+	std::cerr << "\nOutput unit_vectors to ./realsense_unit_vectors.obj!\n\n" << std::endl;
+
+	for (i = 0; i < height; i++)
+		for (j = 0; j < width * 3; j = j + 3)
+			fprintf(fp, "v %.6f %.6f %.6f\n", unitVectors[i][j], unitVectors[i][j + 1], unitVectors[i][j + 2]);
+
+	fclose(fp);
+
+	std::cerr << "\n*** End of unit vectors ***\n\n" << std::endl;
+}
+
+void BaseRealSenseNode::publishUnitVectorsTopic(const ros::Time& t, const std::map<stream_index_pair, bool>& is_frame_arrived)
+{
+	static float num_valid;
+	static bool unit_vector_mature;
+	static unsigned int height = 240, width = 424, sequence;
+	static float current_unit_vectors[FRAME_PRIME_HEIGHT][FRAME_PRIME_WIDTH * 3]= {0};
+
+	try
+	{
+		if (!is_frame_arrived.at(DEPTH))
+		{
+			ROS_DEBUG("Brain: Skipping publish unit vector topic! Depth frame didn't arrive.");
+			return;
+		}
+	}
+	catch (std::out_of_range)
+	{
+		ROS_DEBUG("Brain: Skipping publish unit vector topic! Depth frame didn't configure.");
+		return;
+	}
+
+	auto image_depth16 = reinterpret_cast<const uint16_t*>(_image[DEPTH].data);
+	auto depth_intrinsics = _stream_intrinsics[DEPTH];
+
+	height = depth_intrinsics.height;
+	width = depth_intrinsics.width;
+
+	if ((detailed_logging) && (max_log-- > 0))
+		ROS_INFO("Brain: depth_intrinsics: width:%d; height:%d; _depth_scale_meters:%.6f.\n",
+					width, height, _depth_scale_meters);
+
+	if (!unit_vector_mature) {
+		float scaled_depth, num_bad_pixel = 0;
+
+		for (int y = 0; y < depth_intrinsics.height; ++y)
+		{
+			for (int x = 0; x < depth_intrinsics.width; ++x)
+			{
+				scaled_depth = static_cast<float>(*image_depth16) * _depth_scale_meters;
+
+				++image_depth16;
+
+				if (fabs(scaled_depth) < 0.0001) {
+					num_bad_pixel++;
+					continue;
+				}
+
+				if (valid_pixels[y][x]) {
+					continue;
+				}
+
+				unit_vectors[y][x * 3] = (x - depth_intrinsics.ppx) / depth_intrinsics.fx / scaled_depth;
+				unit_vectors[y][x * 3 + 1] = (y - depth_intrinsics.ppy) / depth_intrinsics.fy / scaled_depth;
```

Comment:
Gonna nitpick on these two lines mainly just for readability, and I'm sure after the compiler's done it makes no difference but... could we just write `(y - depth_intrinsics.ppy) / (depth_intrinsics.fy * scaled_depth)` instead of the double divide?